### PR TITLE
generic netlink protocol support

### PR DESCRIPTION
**Pull request by Leo1003**
This PR adds two crates to add support for generic netlink protocol:
- `netlink-packet-generic`: the packet implementation for generic netlink and its control family
- `genetlink`: the async connection and handle of generic netlink protocol

It is welcome to provide some suggestions for this PR!

### Remaining questions
- None
### Resolved questions
- How to resolve & fill the dynamic family ID?
    - Provide a getter/setter on the message to let it filled by other high level crates

### Dependencies
- another PR(#151)

### CODE REVIEW

**Comment by cathay4t** [comment]
@Leo1003 Sorry for the overlapping. Could take a look on #155 and #157 to see whether we can join the efforts.

**Comment by Leo1003** [comment]
My implement of generic control protocol is not better enough.  
I think it's great if we can integrate the work.  
I will investigate them in a few days, and try to write some documents before merging.

**Comment by Leo1003** [comment]
I have read #155, it's an amazing work!  
I think the packet definition can integrate with this PR, and #155 can serve as a high level crate for generic netlink.  
This should solve the issue of the dynamic family ID in this PR.

**Comment by cathay4t** [comment]
@Leo1003 Could provide example code on using this dynamic family ID, I only see a setter and getter, nothing dynamic? IMHO, the `GenerciNetlink` API does not required to handle family ID auto-setup issue. But the `Ethtool` API should __not__ require use to find out family ID manually, which I failed to find a good way to do so.

**Comment by Leo1003** [comment]
@cathay4t The reason that it only has a getter and a setter is because this crate only serve to define the packet structure. The dynamic resolving should not contain in this. So, I want to make it have a interface for high level crates to set the family ID before serializing.

I also considered define them on the trait, or not to handle at all. But I am not sure which is better, this needs more opinions or examples.

**Comment by Leo1003** [comment]
I am going to write some documentations for the crates. Hope I can complete this and make it ready for review in a few days.

**Comment by Leo1003** [comment]
I think it's ready!

**Comment by Leo1003** [comment]
Whoops... I didn't notice that I forgot to remove "WIP" from the PR title. Thanks.

**Comment by cathay4t** [comment]
@Leo1003 Thanks for the patch. This will take a while for me to review it. During the time, could you include test code as Github Action support netlink in Ubuntu 20.04?

**Comment by Leo1003** [comment]
@cathay4t OK, I would make some tests or examples. I think that test against controller (nlctrl) should be enough, right?

**Review by cathay4t — CHANGES_REQUESTED** [review]
1. Please remove unused code. `Vec<u8>` is not a good place holder, use `struct` if you prefer a place holder over removing unused code.
2. The `Trait GenFamily` is not a good choice for ethtool/wifi/wireguard to implement. Unless you may provide proving code using this PR to implement ethtool/wireguard/wifi/etc.
3. Could you check https://github.com/little-dude/netlink/pull/155 and let me know why patch there not fit your needs? I have put that code in production use(via nmstate over nispor) for months on ethtool(via generic netlink interface). This PR seems a duplication to me.

**Comment by Leo1003** [comment]
>     3. Could you check [Introduce generic netlink(NETLINK_GENERIC) initial support. #155](https://github.com/little-dude/netlink/pull/155) and let me know why patch there not fit your needs? I have put that code in production use(via nmstate over nispor) for months on ethtool(via generic netlink interface). This PR seems a duplication to me.

@cathay4t  I found that this repository uses two crates to implement one netlink protocol. One crate is defining packet structure and handling serialization only. Another one uses `netlink-proto` to provide higher level implementation. 

I prefer to the project structure and planned to create two crates in two PRs at first. During I working on this, #155 implements the higher level one with packet definitions included.

If you think that it's fine to make netlink generic protocol with only one crate. Then we can close this PR and contribute to #155.
Or to merge this PR first, and make #155 to use the packet definition in this crate.

File: netlink-packet-generic/src/message.rs
Comment (Leo1003) [reply netlink-packet-generic/src/message.rs:60]: I think I didn't explain the usage correctly in the doc comment. I would fix it with more detailed documents or an example.

This method is not related to any async operation. It is just a getter/setter functions to let other resolver crates to store resolved family id resolved by themselves. The resolving process should be done by other crates.

In `NetlinkSerializable::message_type()` would return the previous resolved family id store by `set_resolved_family_id()`.

The total process is:
1. user creates the payload
2. user send the payload using other higher level crates
3. higher level crates resolve family id from netlink controller or caches
4. store it into the message using `set_resolved_family_id()`
5. call `GenlMessage::finalize()`
6. Turn into `NetlinkMessage`
7. call `NetlinkMessage::finalize()`
    - In this step, the family id filled into the netlink header automatically!
8. serialize with `NetlinkMessage::serialize()`
9. send it with socket

Comment (cathay4t) [reply netlink-packet-generic/src/message.rs:60]: Can you show me a code?
When I play with ethtool, I have to make two connections with different return message type(one for General netlink, one for ethtool message). I failed to understand how to code a `netlink-ethtool` using this trait.

**Comment by cathay4t** [comment]
> > ```
> > 3. Could you check [Introduce generic netlink(NETLINK_GENERIC) initial support. #155](https://github.com/little-dude/netlink/pull/155) and let me know why patch there not fit your needs? I have put that code in production use(via nmstate over nispor) for months on ethtool(via generic netlink interface). This PR seems a duplication to me.
> > ```
> 
> @cathay4t I found that this repository uses two crates to implement one netlink protocol. One crate is defining packet structure and handling serialization only. Another one uses `netlink-proto` to provide higher level implementation.

The PR #155 does not have code changes to `netlink-proto`. Could you elaborate the `two crates`?

> 
> I prefer to the project structure and planned to create two crates in two PRs at first. During I working on this, #155 implemented the higher level one with packet definitions included.
> 
> If you think that it's fine to make netlink generic protocol support with only one crate. Then we can close this PR and contribute to #155. Or to merge this PR first, and make #155 to use the packet definition in this crate.
> 
> I am sorry that I didn't explain it first.

Without code on ethtool/wireguard/wifi/etc showing this design actuall works, I cannot approve this design. Sorry.

**Comment by Leo1003** [comment]
I initially want to implement a Wireguard library, but the netlink part is not yet completed. 

It is sad that I cannot provide a working design now.

**Comment by Leo1003** [comment]
Well, with a day of copy & paste. I created a rough packet definition of Wireguard in my side project [wiremole/netlink-packet-wireguard](https://github.com/Leo1003/wiremole/tree/master/netlink-packet-wireguard).

Maybe I can make a working design in a few days.

**Comment by cathay4t** [comment]
You example does not have `family_id` implementation which is my biggest confusing of this PR.

**Comment by Leo1003** [comment]
> You example does not have `family_id` implementation which is my biggest confusing of this PR.

Yes, I knew that. It's just a try. 

Therefore, I implemented a high level generic netlink crate called `genetlink` to demonstrate it. There is a family id resolver (with caches!) in it to show that how the `family_id` related functions works.

I also updated the [example](https://github.com/Leo1003/netlink/blob/wireguard/netlink-packet-wireguard/examples/get_wireguard_info.rs) in `netlink-packet-wireguard` to use the new crate. 

P.s. I moved the `netlink-packet-wireguard` into [another branch](https://github.com/Leo1003/netlink/tree/wireguard) in my fork. To make it easy to merge it into this repository in the future.

Hope these can help you answer you questions. :)

**Comment by cathay4t** [comment]
Now I get your point. Thanks for the great work. I have closed #155 in the favor of this design.

Todo for this PR:
 1. squash patches into single one.
 2. Include both `genetlink` and `netlink-packet-generic`.
 

I will try to point my ethtool crate to your design and see how that goes.

**Comment by Leo1003** [comment]
Ok. I will also add documentations and examples & tests for `genetlink` before squashing into one commit.

**Comment by Leo1003** [comment]
@cathay4t I think that I should mention you to review this again.

**Comment by cathay4t** [comment]
@Leo1003 in the first comment, you indicate there are 3 remaindering tasks. Are they all finished or you would do it in other PR.

**Comment by Leo1003** [comment]
@cathay4t The task list is just indicating which method are chosen in the final. Maybe I should not use a task list to represent that?

File: genetlink/Cargo.toml
Comment (cathay4t) [inline genetlink/Cargo.toml:24]: This is new dependency which can be easily replaced by two `HashMap`.
New dependency means more maintenance work for developer and package builder.
Please replace it with HashMap/etc

File: genetlink/src/resolver.rs
Comment (cathay4t) [inline genetlink/src/resolver.rs:145]: Please use `warn!()` instead.

**Review by cathay4t — CHANGES_REQUESTED** [review]
Please remove the use of `bimap` then we are ready to merge.

It could be better If you could also isolate the unrelated decorative changes to standalone patch.

**Comment by cathay4t** [comment]
The main reason I dislike introduce new dependency  is: Dependency addition should be carefully reviewed and only approved when benefit transcends risk.

The main risk of `bimap` is Fedora does not has it packaged, if I want to update this package in fedora, I have to be the packager of `bimap` also which clearly I don't want to be.

File: genetlink/src/resolver.rs
Comment (Leo1003) [reply genetlink/src/resolver.rs:145]: It is inside the unit test, I don't think that it should use `warn!()` instead.
Unless you mean using the following work around:
```rust
#[cfg(test)]
use std::{println as info, println as warn};
```

File: genetlink/Cargo.toml
Comment (Leo1003) [reply genetlink/Cargo.toml:24]: The usage of bimap is removed.

File: genetlink/src/resolver.rs
Comment (cathay4t) [reply genetlink/src/resolver.rs:145]: Oh. I missed the unit test part. Yes, you can use print in unit test.

**Comment by cathay4t** [comment]
@Leo1003 I still see bimap in your latest commit. Could you check again?

**Comment by Leo1003** [comment]
I accidentally overwrote the commit when I rebased the PR onto master.
It is fixed now.
Sorry for that.

**Review by cathay4t — APPROVED** [review]
LGTM. Thanks

**Comment by cathay4t** [comment]
@Leo1003 Patch merged. I will work on the ethtool part. Looking forward to your wireguard crate. Thanks fro the great work!

**Pull request merged**
